swap/simulations_test.go

```diff
@@ -0,0 +1,752 @@
+			ts.spec.Hook = protocols.NewAccounting(balance)
+			ts.swap = balance
+			// deploy the accounting to the `SimulatedBackend`
+			err = testDeploy(context.Background(), balance)
+			if err != nil {
+				return nil, nil, err
+			}
+			// store the testService into the bucket
+			bucket.Store(bucketKeySwap, ts)
+
+			cleanup = func() {
+				ts.swap.store.Close()
+				os.RemoveAll(dir)
+			}
+
+			return ts, cleanup, nil
+		},
+	}
+	return simServiceMap
+}
+
+// newSharedBackendSwaps pre-loads each simulated node account with "funds"
+// so that later in the simulation all operations have sufficient gas
+func newSharedBackendSwaps(nodeCount int) (*swapSimulationParams, error) {
+	params := &swapSimulationParams{
+		swaps:       make(map[int]*Swap),
+		dirs:        make(map[int]string),
+		maxMsgPrice: 10000,
+		minMsgPrice: 100,
+		nodeCount:   nodeCount,
+	}
+	keys := make(map[int]*ecdsa.PrivateKey)
+	addrs := make(map[int]common.Address)
+	alloc := core.GenesisAlloc{}
+	stores := make(map[int]*state.DBStore)
+
+	// for each node, generate keys, a GenesisAccount and a state store
+	for i := 0; i < nodeCount; i++ {
+		key, err := crypto.GenerateKey()
+		if err != nil {
+			return nil, err
+		}
+		keys[i] = key
+		addrs[i] = crypto.PubkeyToAddress(key.PublicKey)
+		alloc[addrs[i]] = core.GenesisAccount{Balance: big.NewInt(10000000000)}
+		dir, err := ioutil.TempDir("", fmt.Sprintf("swap_test_store_%x", addrs[i].Hex()))
+		if err != nil {
+			return nil, err
+		}
+		stateStore, err2 := state.NewDBStore(dir)
+		if err2 != nil {
+			return nil, err
+		}
+		params.dirs[i] = dir
+		stores[i] = stateStore
+	}
+	// then create the single SimulatedBackend
+	gasLimit := uint64(8000000000)
+	defaultBackend := backends.NewSimulatedBackend(alloc, gasLimit)
+	testBackend := &swapTestBackend{SimulatedBackend: defaultBackend}
+	// finally, create all Swap instances for each node, which share the same backend
+	for i := 0; i < nodeCount; i++ {
+		params.swaps[i] = new("", stores[i], keys[i], testBackend, DefaultDisconnectThreshold, DefaultPaymentThreshold)
+	}
+
+	params.backend = testBackend
+	return params, nil
+}
+
+// TestPingPongChequeSimulation just launches two nodes and sends each other
+// messages which immediately crosses the PaymentThreshold and triggers cheques
+// to each other. Checks that accounting and cheque handling works across multiple
+// cheques and also if cheques are mutually sent
+func TestPingPongChequeSimulation(t *testing.T) {
+	nodeCount := 2
+	// create the shared backend and params
+	params, err := newSharedBackendSwaps(nodeCount)
+	if err != nil {
+		t.Fatal(err)
+	}
+	// cleanup backend
+	defer params.backend.Close()
+
+	// setup the wait for mined transaction function for testing
+	cleanup := setupContractTest()
+	defer cleanup()
+
+	params.backend.cashDone = make(chan struct{}, 1)
+	defer close(params.backend.cashDone)
+	// initialize the simulation
+	sim := simulation.NewInProc(newSimServiceMap(params))
```

Comment:
consider `NewBzzInProc` since it handles some of the boilerplate for the simulation